lib/puppet/generate/type.rb

```diff
@@ -46,7 +46,24 @@ def format=(format)
         # @return [Boolean] Returns true if the output is up-to-date or false if not.
         def up_to_date?(outputdir)
           f = effective_output_path(outputdir)
-          Puppet::FileSystem::exist?(f) && (Puppet::FileSystem::stat(@path) <=> Puppet::FileSystem::stat(f)) <= 0
+          # Check the fast-path scenarios first.
+          unless Puppet::FileSystem::exist?(f)
+            Puppet.debug("#{f} does not exist.")
+            return false
+          end
+          unless (Puppet::FileSystem::stat(@path) <=> Puppet::FileSystem::stat(f)) <= 0
+            Puppet.debug("#{@path} is newer than #{f}")
+            return false
+          end
+          # Check for updates to any module lib files.
+          module_lib_files = Dir.glob(File.join(@base, "lib", "**", "*.rb"))
+          module_lib_files.each do |lib|
```

Comment:
Wondering if it'd be better to use the block form and break as soon as we detect a newer file?
```suggestion
Dir.glob(File.join(@base, "lib", "**", "*.rb")) do |lib|
  ...
end
 ```

Also I'd want to understand how this would affect code deployment times for a "typical" environment with hundreds of modules.